rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,364 @@
+7. The desired agent version cannot be set via Teleport's operator-targeted CLI (tctl).
+8. The rollout plan for the new agent version is not fully-configurable using tctl.
+9. Agent installation logic is spread between the auto-updater script, install script, auto-discovery script, and documentation.
+10. Teleport contains logic that is specific to Teleport Cloud upgrade workflows.
+11. The existing auto-updater is not self-updating.
+12. It is difficult and undocumented to automate agent upgrades with custom automation (e.g., with JamF). 
+
+We must provide a seamless, hands-off experience for auto-updates that is easy to maintain.
+
+## Details
+
+We will ship a new auto-updater package written in Go that does not interface with the system package manager.
+It will be versioned separately from Teleport, and manage the installation of the correct Teleport agent version manually. 
+It will read the unauthenticated `/v1/webapi/ping` endpoint from the Teleport proxy, parse new fields on that endpoint, and install the specified agent version according to the specified upgrade plan.
+It will download the correct version of Teleport as a tarball, unpack it in `/var/lib/teleport`, and ensure it is symlinked from `/usr/local/bin`.
+
+### Installation
+
+```shell
+$ apt-get install teleport-ent-updater
+$ teleport-update enable --proxy example.teleport.sh
+
+# if not enabled already, configure teleport and:
+$ systemctl enable teleport
+```
+
+### API
+
+#### Endpoints
+
+`/v1/webapi/ping`
+```json
+{
+  "server_edition": "enterprise",
+  "agent_version": "15.1.1",
+  "agent_auto_update": true,
+  "agent_update_after": "2024-04-23T18:00:00.000Z",
+  "agent_update_jitter_seconds": 10,
+}
+```
+Notes:
+- Critical updates are achieved by serving `agent_update_after` with the current time.
+- The Teleport proxy translates upgrade hours (below) into a specific time after which all agents should be upgraded.
+- If an agent misses an upgrade window, it will always update immediately.
+- The edition served is the cluster edition (enterprise, enterprise-fips, or oss), and cannot be configured.
+
+#### Teleport Resources
+
+```yaml
+kind: cluster_maintenance_config
+spec:
+  # agent_auto_update allows turning agent updates on or off at the
+  # cluster level. Only turn agent automatic updates off if self-managed
+  # agent updates are in place.
+  agent_auto_update: on|off
+  # agent_update_hour sets the hour in UTC at which clients should update their agents.
+  agent_update_hour: 0-23
+  # agent_update_now overrides agent_update_hour and sets agent update time to the current time.
+  # This is useful for rolling out critical security updates and bug fixes.
+  agent_update_now: on|off
+  # agent_update_jitter_seconds sets a duration in which the upgrade will occur after the hour.
+  # The agent upgrader will pick a random time within this duration in which to upgrade.
+  agent_update_jitter_seconds: 0-MAXINT64
```

Comment:
Will updates run daily? The jitter value should be capped at `Update period - max duration of an update action`.

Reply:
There is no fixed cadence for agent upgrades. The systemd timer runs `teleport-updater update` every 10 min. It could be configured to avoid concurrent executions, but also happy to limit jitter to 300s (half the interval) if you think that jitter duration longer than the polling interval creates too much risk that upgrades would be delayed longer than intended.